pandas/tseries/offsets.py

```diff
@@ -603,61 +641,76 @@ def next_bday(self):
+                                latest_start.hour, latest_start.minute)
         else:
-            if self.n >= 0 and self.start < other.time():
-                other = other + self.next_bday
-            elif self.n < 0 and other.time() < self.start:
-                other = other + self.next_bday
-        return datetime(other.year, other.month, other.day,
-                        self.start.hour, self.start.minute)
+            if self.n * sign >= 0 and latest_start < other.time():
+                # current time is after latest starting time in today
+                other = other + sign * self.next_bday
+                return datetime(other.year, other.month, other.day,
+                                earliest_start.hour, earliest_start.minute)
+            elif self.n * sign < 0 and other.time() < earliest_start:
+                # current time is before earliest starting time in today
+                other = other + sign * self.next_bday
+                return datetime(other.year, other.month, other.day,
+                                latest_start.hour, latest_start.minute)
+        if self.n * sign >= 0:
+            # find earliest starting time later than or equal to current time
+            for st in self.start:
+                if other.time() <= st:
+                    return datetime(other.year, other.month, other.day,
+                                    st.hour, st.minute)
+        else:
+            # find latest starting time earlier than or equal to current time
+            for st in reversed(self.start):
+                if other.time() >= st:
+                    return datetime(other.year, other.month, other.day,
+                                    st.hour, st.minute)
 
     def _prev_opening_time(self, other):
         """
-        If n is positive, return yesterday's business day opening time.
-        Otherwise yesterday business day's opening time.
+        If n is positive, return the latest opening time earlier than or equal
+        to current time.
+        Otherwise the earliest opening time later than or equal to current
+        time.
+
         """
-        if not self.next_bday.onOffset(other):
-            other = other - self.next_bday
-        else:
-            if self.n >= 0 and other.time() < self.start:
-                other = other - self.next_bday
-            elif self.n < 0 and other.time() > self.start:
-                other = other - self.next_bday
-        return datetime(other.year, other.month, other.day,
-                        self.start.hour, self.start.minute)
+        return self._next_opening_time(other, sign=-1)
 
-    @cache_readonly
-    def _get_business_hours_by_sec(self):
+    def _get_business_hours_by_sec(self, start, end):
         """
         Return business hours in a day by seconds.
         """
-        if self._get_daytime_flag:
+        if self._get_daytime_flag(start, end):
             # create dummy datetime to calculate businesshours in a day
```

Comment:
do same as above, only define hour, minute in the clauses

Reply:
Updated